clock_exercise/alarmitemmodel.cpp

```diff
@@ -0,0 +1,208 @@
+        case MonthRole:
+            return m_data[row].m_month;
+        case YearRole:
+            return m_data[row].m_year;
+        case HoursRole:
+            return m_data[row].m_hours;
+        case MinutesRole:
+            return m_data[row].m_minutes;
+    }
+
+    return QVariant();
+}
+
+QHash<int, QByteArray> AlarmItemModel::roleNames() const
+{
+    return m_roleNames;
+}
+
+void AlarmItemModel::setAllNotSelected() {
+    for(int i = 0; i < m_data.count(); i++) {
+        if(m_data[i].m_selected == true) {
+            m_data[i].m_selected = false;
+            emit dataChanged( QAbstractListModel::index(i),  QAbstractListModel::index(i));
+        }
+    }
+    setNumberSelectedAlarms(0);
+}
+
+bool AlarmItemModel::setData(const QModelIndex& index, const QVariant& value, int role)
+{
+    int row = index.row();
+    if(row < 0 || row >= m_data.count()) {
+        return false;
+    }
+
+    switch (role) {
+    case EverydayRole:
+        if (value != m_data[row].m_everyday) {
+            m_data[row].m_everyday = value.toBool();
+            emit dataChanged(index, index);
+            return true;
+        }
+    case ActiveRole:
+        if (value != m_data[row].m_active) {
+            m_data[row].m_active = value.toBool();
+            emit dataChanged(index, index);
+            if(m_data[row].m_active)
+                setNumberActiveAlarms(m_numberActiveAlarms + 1);
+            else
+                setNumberActiveAlarms(m_numberActiveAlarms - 1);
+            return true;
+        }
+    case SelectedRole:
+        if (value != m_data[row].m_selected) {
+            m_data[row].m_selected = value.toBool();
+            emit dataChanged(index, index);
+            if(m_data[row].m_selected)
+                setNumberSelectedAlarms(m_numberSelectedAlarms + 1);
+            else
+                setNumberSelectedAlarms(m_numberSelectedAlarms - 1);
+            return true;
+        }
+    case DayRole:
+        if (value != m_data[row].m_day) {
+            m_data[row].m_day = value.toInt();
+            emit dataChanged(index, index);
+            return true;
+        }
+    case MonthRole:
+        if (value != m_data[row].m_month) {
+            m_data[row].m_month = value.toInt();
+            emit dataChanged(index, index);
+            return true;
+        }
+    case YearRole:
+        if (value != m_data[row].m_year) {
+            m_data[row].m_year = value.toInt();
+            emit dataChanged(index, index);
+            return true;
+        }
+    case HoursRole:
+        if (value != m_data[row].m_hours) {
+            m_data[row].m_hours = value.toInt();
+            emit dataChanged(index, index);
+            return true;
+        }
+    case MinutesRole:
+        if (value != m_data[row].m_minutes) {
+            m_data[row].m_minutes = value.toInt();
+            emit dataChanged(index, index);
+            return true;
+        }
+        break;
+    }
+
+    return false;
```

Comment:
Non sono sicuro che si debba ritornare `false` in tutti quei casi in cui il valore non sia cambiato.
Di fatto se io cerco di impostare a `4` il valore di una property e quella property ha già quel valore, direi che l'operazione ha comunque avuto successo indipendentemente dal fatto che non sia cambiata.
La documentazione di Qt non è estremamente chiara sul significato di `successful` (https://doc.qt.io/qt-5/qabstractitemmodel.html#setData), per cui chiederei a Luca su questo punto.

Sicuramente però aggiungerei un case `default` allo switch e da lì ritornerei `false`.

Reply:
Sono d'accordo, ma non sono d'accordo con la soluzione, mettendo il `return false` nel default è identico ad adesso. Credo che vada spostato il `return true` al di fuori di  `if (value != m_data[row].m_month)`, ma comunque dentro al `case MonthRole:`